designs/2024-support-ts-config-files/README.md

```diff
@@ -0,0 +1,132 @@
+- Repo: eslint/eslint
+- Start Date: 2024-03-09
+- RFC PR: <https://github.com/eslint/rfcs/pull/117>
+- Authors: [Arya Emami](https://github.com/aryaemami59)
+
+# Add Support for TypeScript Config Files
+
+## Summary
+
+Add support for TypeScript config files (`eslint.config.ts`, `eslint.config.mts`, `eslint.config.cts`)
+
+<!-- One-paragraph explanation of the feature. -->
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+The primary motivation for adding support for TypeScript configuration files to ESLint is to enhance the developer experience and accommodate the evolving JavaScript ecosystem. As TypeScript's popularity continues to grow, more projects are adopting TypeScript not only for their source code but also for their configuration files. This shift is driven by TypeScript's ability to provide compile-time type checks and IntelliSense. By supporting `eslint.config.ts`, `eslint.config.mts`, and `eslint.config.cts`, ESLint will offer first-class support to TypeScript users, allowing them to leverage these benefits directly within their ESLint configuration.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+The goal is to seamlessly support TypeScript configuration files in ESLint. To achieve this, ESLint will need to recognize and parse TypeScript configuration files in the same way it does for JavaScript configuration files. This will involve creating the configuration file resolution logic to recognize `.ts`, `.mts`, and `.cts` files as valid ESLint configuration files. We will need to treat `eslint.config.mts` files as ECMAScript Modules (ESM) and `eslint.config.cts` files as CommonJS (CJS). The module resolution for `eslint.config.ts` files will depend on the `type` field in the nearest `package.json`. The maintainers of ESLint have raised some valid concerns some of which include:
```

Comment:
Can you give an example of how eslint configs should be typed?  as of now, eslint does not export typings.

tools like vite support an api: `defineConfig()`, should we do the similar thing?

```ts
import "defineConfig" from "eslint";

export default defineConfig([
....
]);
```

Reply:
I would absolutely love that, but that's up to the maintainers, I would be happy to include a `defineConfig()` or `createESLintConfig()` in the linked PR or even in a separate PR. All I need is the thumbs up!

Reply:
Just for reference, I have something like that setup in [Redux-Toolkit](https://github.com/reduxjs/redux-toolkit/blob/94fb3a77f95f8f10e939c32d7a2050511a786bdd/packages/toolkit/eslint.config.mjs).

In an ESM file it can look like this:

```ts
import { createESLintConfig } from '@reduxjs/eslint-config'

export default createESLintConfig([
  {
    rules: {
      'no-console': [0],
    },
  },
  {
    // ...Other additional overrides
  },
])
```

and in a CJS file it can look like this:

```ts
module.exports = (async () =>
  (await import('@reduxjs/eslint-config')).createESLintConfig([
    {
      rules: {
        'no-console': [0],
      },
    },
    {
      // ...Other additional overrides
    },
  ]))()
```

Reply:
yes, it's worthy to be listed in the open questions.

Reply:
Done!